include/bio/plain_io/writer.hpp

```diff
@@ -0,0 +1,633 @@
+ *
+ * Line-wise writing:
+ * ```cpp
+ * bio::plain_io::writer writer{"example.txt"};
+ *
+ * writer.push_back("line 1");
+ * writer.emplace_back("line 2 contains the integer ", 3, " and the float ", 4.3, "!");
+ * writer.push_back("line 3");
+ * ```
+ *
+ * Column-wise writing:
+ *```cpp
+ * bio::plain_io::writer writer{"example.tsv.gz", '\t'}; // transparent compression supported
+ *
+ * writer.push_back("This is the header line");
+ * writer.emplace_back("col1", 3.4, "col3", 4);          // will be printed tab-separated
+ *```
+ */
+template <record_kind record_kind_>
+class writer
+{
+private:
+    //!\brief The element type of this range.
+    using value_type = std::conditional_t<record_kind_ == record_kind::line, std::string_view, record>;
+
+public:
+    /*!\name Range associated types
+     * \brief The types necessary to facilitate the behaviour of an input range (used in record-wise reading).
+     * \{
+     */
+    //!\brief The iterator type of this view (an input iterator).
+    using iterator       = detail::plaintext_output_iterator<record_kind_>;
+    //!\brief The const iterator type is void, because files are not const-iterable.
+    using const_iterator = void;
+    //!\brief The type returned by end().
+    using sentinel       = std::default_sentinel_t;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    //!\brief Default constructor is explicitly deleted, you need to give a stream or file name.
+    writer()               = delete;
+    //!\brief Copy construction is explicitly deleted, because you can't have multiple access to the same file.
+    writer(writer const &) = delete;
+    //!\brief Move construction is defaulted.
+    writer(writer &&)      = default;
+    //!\brief Copy assignment is explicitly deleted, because you can't have multiple access to the same file.
+    writer & operator=(writer const &) = delete;
+    //!\brief Move assignment is defaulted.
+    writer & operator=(writer &&) = default;
+    //!\brief Destructor is defaulted.
+    ~writer()                     = default;
+
+    /*!\brief Construct from filename.
+     * \param[in] filename        Path that you wish to write to.
+     * \param[in] field_separator Delimiter between fields in a line. [optional]
+     * \param[in] ostream_options Options passed to the underlying stream; see seqan3::transparent_ostream_options.
+     * [optional] \throws seqan3::file_open_error If the file could not be opened, e.g. non-existant or non-readable.
+     *
+     * \details
+     *
+     * TODO example
+     *
+     * ### Compression
+     *
+     * This constructor transparently applies a compression stream on top of the file stream in case
+     * the filename has an extension that indicates that. You can explicitly request a specific kind of compression
+     * via the ostream_options.
+     */
+    writer(std::filesystem::path const &       filename,
+           char const                          field_separator,
```

Comment:
So many constructors :scream_cat: 

So you set the field separator at construction but what about the `add_CR` variable? What if you decide at some point another option might be handy? It might be worth it to also have a options struct no?

Reply:
The field separator being separate is the only somewhat intuitive way to select between line-only and line-and-fields type; this happens via deduction guides. It looks like this:
```cpp
bio::plain_io::reader r{"foo.txt"}; // lines
bio::plain_io::reader r{"foo.txt", '\t'}; // lines-and-fields
```
I will add snippets for this.

If we do this via options, we (a) get a default that isn't used for line-only-reading (which is confusing) and (b) we need to make those options a template and add some tagging mechanism to (de-)activate line-and-fields.
It would look something like this:
```cpp
bio::plain_io::reader r{"foo.txt"}; // lines
bio::plain_io::reader r{"foo.txt", bio::plain_io::reader_options{.field_seperator = '\t', .split_lines = seqan3::vtag<true>}; // lines-and-fields
```
 I think that this is unnecessary complexity for something that should "plain". 

add_CR is something that I really don't like and -- as the design note states -- should be difficult to set.

Reply:
yes I agree that the former is nicer. I  guess it stays like this until another option pops up. We can always have an option contructor side by side with these.

Reply:
If we really find more options, we can think about changing this, but if we can avoid it, I would prefer that.

Reply:
Also `add_CR` is only available in the writer, so we would have to make different options for reader and writer...